plugin-server/src/worker/ingestion/event-pipeline/runner.ts

```diff
@@ -117,22 +118,59 @@ export class EventPipelineRunner {
             // ingestion pipeline is working well for all teams.
             this.poEEmbraceJoin = true
         }
-        const processedEvent = await this.runStep(pluginsProcessEventStep, [this, event], event.team_id)
 
+        let processPerson = true
+        if (event.properties && event.properties.$process_person === false) {
+            // We are purposefully being very explicit here. The `$process_person` property *must*
+            // exist and be set to `false` (not missing, or null, or any other value) to disable
+            // person processing.
+            processPerson = false
+
+            if (['$identify', '$create_alias', '$merge_dangerously', '$groupidentify'].includes(event.event)) {
+                const warningAck = captureIngestionWarning(
+                    this.hub.db.kafkaProducer,
+                    event.team_id,
+                    'invalid_event_when_process_person_is_false',
+                    {
+                        eventUuid: event.uuid,
+                        event: event.event,
+                        distinctId: event.distinct_id,
+                    },
+                    { alwaysSend: true }
+                )
+
+                return this.registerLastStep('invalidEventForProvidedFlags', [event], [warningAck])
+            }
+        }
+
+        const processedEvent = await this.runStep(pluginsProcessEventStep, [this, event], event.team_id)
         if (processedEvent == null) {
+            // A plugin dropped the event.
             return this.registerLastStep('pluginsProcessEventStep', [event])
         }
-        const [normalizedEvent, person] = await this.runStep(processPersonsStep, [this, processedEvent], event.team_id)
 
-        const preparedEvent = await this.runStep(prepareEventStep, [this, normalizedEvent], event.team_id)
+        // Normalizing is sync and doesn't need to run in a full `runStep` span for tracking.
+        const [normalizedEvent, timestamp] = normalizeEventStep(processedEvent, processPerson)
```

Comment:
the step stuff also does some error handling, do we need that (see related question in the new file)?

Reply:
https://github.com/PostHog/posthog/pull/21262#discussion_r1552276528